17 use GitHub token

gbataille opened the pull request:
`--github-key` as a command line argument to pass a personal github token. 

Circumvents the rate limit issue and gives access to private repo (providing the token has the right scopes)

gbataille committed:
feat: github-key as commandline param

gbataille committed:
refactor: rename sample -> optionsParser

gbataille committed:
feat: use github token for API calls

solves #17
solves #18

guibou commented on krank.cabal:
```diff
@@ -31,7 +31,8 @@ library
                        , unordered-containers >= 0.2.10 && < 0.2.11
-  hs-source-dirs:      src
+                       , utf8-string >= 1.0.1.1 && < 1.1
+  Hs-source-dirs:      src
```
Why the capitalization here?

gbataille replied:
because it looks nicer? :)

I must have done that without realizing it

guibou commented on app/Main.hs:
```diff
@@ -3,32 +3,40 @@
 data KrankOpts = KrankOpts {
-  codeFilePaths :: [FilePath]
+  codeFilePaths :: [FilePath],
+  githubKey :: Maybe String
```
Could you wrap `githubKey` in a `newtype`?

gbataille replied:
Following our offline discussion, I did a `newtype` wrap to have a clean function signature 
--> is that what you wanted?

Actually, why `newtype`? My (maybe wrong understanding) is that to do that `data` and `newtype` will do the same but `data` is preferred and `newtype` is rather used when we specifically want a "redefinition" that will implement differently some typeclasses. Am I missing something?

guibou replied:
Thank you for the change. Actually I was expectding a `Maybe GithubKey` with `GithubKey` just a `newtype` over `String`, but I have no argument for or against your solution.

`newtype` versus `data` is really simple. `newtype` comes for free, no overhead and the same behavior of the wrapped type. Especially regarding strictness:

The ctor for a `data` can be undefined:

```haskell
Prelude> data Poulet = Poulet Int
Prelude> foo (Poulet _) = "Ypipp"

Prelude> foo undefined
"*** Exception: Prelude.undefined
CallStack (from HasCallStack):
  error, called at libraries/base/GHC/Err.hs:78:14 in base:GHC.Err
  undefined, called at <interactive>:3:5 in interactive:Ghci3
Prelude> foo (Poulet undefined)
"Ypipp"
```

In this case, Haskell will first evaluate the first "thunk" to see if there is a constructor inside it, and fail on the `undefined`.

On the other hand:

```haskell
Prelude> newtype Poulet = Poulet Int
Prelude> foo (Poulet _) = "Ypipp"


Prelude> foo undefined
"Ypipp"
```

It behaves exactly as an Int, without wrapping constructor.


As for the reasons to prefer one or the other in this context, there is no reason, we don't care about performance neither strictness properties. That's just an habit I have.

guibou reviewed (commented):
Hello!

A big thank you for this PR.

Could you fix the few comments, and especially wrapping the github key in a `newtype`, I'm not motivated by the idea of having a `String` floating somewhere which is in fact important credentials.

One question with this approach. The credential will appear in the command line used to run `krank`, so it will appears in the output of the `ps` tool, so it can be "stolen" by other users of the machine. Am I parano? What is the state of the art relative to security token and command line?

gbataille committed:
refactor: specific type for top level function signature

gbataille commented:
For the token, you are kind of right.
I would prefer to have it in a config file or as an environment variable but
- config file might be accessible too. At some point this "secret" needs to be somewhere.
- Supposedly this token should
  - (for purely public users) have 0 rights and therefore it is used only to get a higher API rate limit
  - (for private repo users) have `repo:status` permission which should give access to the repo info but not the code

So this should be safe enough
(we should document at some point what kind of token is needed)

guibou commented:
Approved. Could you write a small documentation for the generation of github token?

gbataille committed:
refactor: GithubKey Maybe String -> Maybe GithubKey String

gbataille committed:
refactor: rename CLI option for checker namespacing

gbataille committed:
doc: IssueTracker documentation

gbataille commented:
- `Maybe` out of the `newtype` (I prefer it that way too)
- renamed the `github-key` option to `issuetracker-githubkey` to namespace the option with the checker it is for (in the future, will be in a hierarchical configuration file that should make this more user friendly)
- Added an **IssueTracker** documentation
